### PR TITLE
Site Editor: Add rename and trash actions to page panel

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related:
- https://github.com/WordPress/gutenberg/pull/54648
- https://github.com/WordPress/gutenberg/issues/52763
- https://github.com/WordPress/gutenberg/pull/60230
- https://github.com/WordPress/gutenberg/pull/60231

## What?

Alternative to https://github.com/WordPress/gutenberg/pull/54648.

Props to @SaxonF as this PR is mostly just an updated version of https://github.com/WordPress/gutenberg/pull/54648.

## Why?

https://github.com/WordPress/gutenberg/issues/52763

If a page has not Title block there is no means to update the title within the site editor.

## How?

- Refactors the existing `RenameMenuItem` component for templates to a shared location and support pages
- Some tweaks to pass `page` records rather than post id/type (happy to hear ideas on better approaches)
- Adds new actions to the Site Editor's `PostCardPanel`

## Testing Instructions

1. Navigate to Appearance > Editor > Templates
2. Check that the renaming of templates still functions correctly
3. Navigate to the Site Editor's Pages page
4. Select a page and click its preview to edit it
5. Confirm there are rename and trash actions available within the Page Panel
6. Test renaming works
7. Navigate back to the the Page details screen (e.g. via top left icon)
8. Confirm the page has rename and trash actions available in the navigation screen
9. Check the renaming works here as well
10. Via either the details screen or the Page Panel while editing, confirm the trash option works

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/60436221/ed973d79-2616-4de3-9af2-5d53e6e31431

### CODE REVIEW

**Review by jameskoster — COMMENTED** [review]
It's working for me 👍 

As a small tweak could we align the icon with others in the Inspector?

Screenshot 2024-03-27 at 11 51 56

Probably separate, but should we use [DropdownMenu V2](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2--docs)?

**Comment by noisysocks** [comment]
Notes from testing:

- I noticed that _Move to trash_ appears in the detail screen or sidebar dropdown even when the post is already in the trash.
- It would be nice if the text field in the Rename dialog was focused automatically or was the first item in the tab order so that it's quick to rename something with the keyboard.
- If you rename a post/page and don't type in a different name it still pops up with a _Name updated_ toast notification.

I'll look at the code now 👀

File: packages/edit-site/src/components/page-actions/trash-page-menu-item.js
Comment (noisysocks) [inline packages/edit-site/src/components/page-actions/trash-page-menu-item.js:21]: When is `page.title` not an object? I wish `@wordpress/data` handled this for us.

File: packages/edit-site/src/components/page-actions/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/page-actions/index.js]: `item` doesn't specifically describe the value being passed. I'd go with `post` which is the technical name for the entity record. I'd also rename `RenameMenuItem` to `RenamePostMenuItem` if it's not too late to do that.

```js
<RenamePostMenuItem post={ post } onClose={ onClose } />
```

**Review by noisysocks — COMMENTED** [review]
Looking good.

File: packages/edit-site/src/components/page-actions/trash-page-menu-item.js
Comment (Mamaduka) [reply packages/edit-site/src/components/page-actions/trash-page-menu-item.js:21]: > When is page.title not an object?

When consuming an edited record, [raw properties](https://github.com/WordPress/gutenberg/blob/4cfaab6302c862ce264247ebef31b45729f04343/packages/core-data/src/entities.js#L21) will be strings. It also depends on the REST API schema, but it should be the same for post types.

> I wish @wordpress/data handled this for us.

It sort of does, but I always have a problem remembering which selector returns which data shape 😓

**Review by kevin940726 — APPROVED** [review]
I think this is good enough and we can address further feedback in follow-ups!

Apart from the issues @noisysocks mentioned above, I also noticed a potential enhancement:

After closing the renaming modal, it would be nice to be able to restore focus to the actions menu button instead of losing the focus to `body`.

File: packages/edit-site/src/components/page-actions/trash-page-menu-item.js
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/page-actions/trash-page-menu-item.js:21]: Good question 👍 

As George notes, this should now always be getting the edited record so the `title` value should be a string. I'll update this PR accordingly.

**Comment by aaronrobertshaw** [comment]
Appreciate all the reviews and feedback, thank you 👍 

> I noticed that Move to trash appears in the detail screen or sidebar dropdown even when the post is already in the trash.

This came up indirectly on a related PR adding the rename action to the page dataviews. The result was https://github.com/WordPress/gutenberg/pull/60236 to prevent navigation to pages that have been moved to the trash. I'll try and tighten that up further while addressing the other improvements here, like the focus management etc.

File: packages/edit-site/src/components/page-actions/trash-page-menu-item.js
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/page-actions/trash-page-menu-item.js:21]: I was mistaken, it appears that the sidebar nav screen for pages is [not using the edited record](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/sidebar-navigation-screen-page/index.js?plain=1#L41).

For the moment, this trash menu item is getting a record with a string value when accessed via the page panel but an object value in the site editor's sidebar nav. I'd like to explore consistently using the edited record in a follow-up.

**Comment by aaronrobertshaw** [comment]
I think I've addressed all the feedback that hasn't already been marked for a [future follow-up](https://github.com/WordPress/gutenberg/pull/60232#discussion_r1547339556) or [in progress in a separate PR](https://github.com/WordPress/gutenberg/pull/60232#issuecomment-2031295696).

Latest changes include:
- Renaming `item` prop to `post` 
- Updating the new rename component to `RenamePostMenuItem`
- Preventing the rename from happening when the name hasn't changed
- Making the page action panel's button icon align with others in the sidebar
- Focusing the text field automatically when the rename modal renders
- Leaving the page actions dropdown open so focus is returned then when rename modal form is submitted

> Probably separate, but should we use [DropdownMenu V2](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2--docs)?

Agreed. Switching the existing use of the V1 Dropdown to the latest is best done in a separate PR.

Let me know if I've missed anything or there are further improvements we can make.

**Comment by jameskoster** [comment]
Unsure if it's related to this PR, but I see a double snackbar:

Screenshot 2024-04-03 at 10 55 09

**Comment by aaronrobertshaw** [comment]
> Unsure if it's related to this PR, but I see a double snackbar:

I think it is only related to this PR in that it is easier to trigger the situation where the notice appears in two places. 

The behaviour happens on trunk as well when a notice is displayed and there is both a preview and DataView in `list` layout.

At a glance, it looks like the snackbar lists are added to the site editor's:
- [interface skeleton](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/editor/index.js?plain=1#L216)
- [`Page` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/page/index.js?plain=1#L39)

On trunk it can be replicated by:
1. Navigate to Appearance > Editor > Pages 
2. Select page and click preview to edit
3. Update and save page
4. While notice is displayed navigate back to Site Editor's Pages view
5. Note the multiple notices

https://github.com/WordPress/gutenberg/assets/60436221/3c379edd-e22f-457b-bb9c-ad547345115c

I'm not sure what the best solution is here. Perhaps we can conditionally render one of the snackbar lists. At this stage, I think that should be a follow-up to this PR rather than a blocker. What do you think @jameskoster?

**Comment by jameskoster** [comment]
Agree, since the issue also exists on trunk it doesn't need to hold up this work.

**Review by talldan — APPROVED** [review]
LGTM, though I am a little unfamiliar with this part of the codebase.

An observation is that the `RenamePost`, `TrashPage` naming feels a little inconsistent, though perhaps it's because the page component only works for pages (I haven't checked).

I also noticed that Rename is missing in the grid/table views, so that might be nice to add, though I don't know if it was intended to be in the scope of this PR or the related issue, it could be a seperate PR:
Screenshot 2024-04-05 at 1 30 40 pm

**Comment by aaronrobertshaw** [comment]
Thanks for the review @talldan 👍 

> An observation is that the RenamePost, TrashPage naming feels a little inconsistent, though perhaps it's because the page component only works for pages (I haven't checked).

That's correct. The `RenamePost` component can rename multiple post entity types and at present, the Page and Template Delete/Move to Trash are separate. The labelling of those options was also flagged for a future audit as a follow-up.

> I also noticed that Rename is missing in the grid/table views, so that might be nice to add, though I don't know if it was intended to be in the scope of this PR or the related issue, it could be a seperate PR:

This was done via a separate PR (https://github.com/WordPress/gutenberg/pull/60230) and linked in the PR description. It's already been merged to trunk.

I'll get this merged as well and address anything else as a follow-up.

**Comment by youknowriad** [comment]
Hello folks, just a note that we've been working on unifying the inspector panels and details panels everywhere between post and site editors and this PR kind of goes in the opposite direction. Can we please add these to the post editor too.

I'd like to request that any feature developed from now on is added to both or none. Thank you.

**Comment by youknowriad** [comment]
cc @jorgefilipecosta as he was working on something very related to this PR as well.

**Comment by aaronrobertshaw** [comment]
> Can we please add these to the post editor too.

Absolutely! 

Thanks for the nudge @youknowriad, I can follow up on the post-editor side of this next week and unify shared components etc as needed.

> I'd like to request that any feature developed from now on is added to both or none. Thank you.

👍 My apologies, I had a little tunnel vision following the [original PR](https://github.com/WordPress/gutenberg/pull/54648) and [issue](https://github.com/WordPress/gutenberg/issues/52763).

**Comment by youknowriad** [comment]
No worries :) Follow-up sounds great. I know that we're kind of "in the middle", so it's not entirely clear yet to everyone, so it's the right time to share that mindset between all contributors.

**Comment by jorgefilipecosta** [comment]
> cc @jorgefilipecosta as he was working on something very related to this PR as well.

Yes as @youknowriad referred the PR https://github.com/WordPress/gutenberg/pull/60486 which was extracted from an old one at https://github.com/WordPress/gutenberg/pull/59849, also ends up implementing this feature. It basically unifies all the actions across dataviews, left details panel, and inspector panel of site editor, edit post. I guess after it is merged actions will be available in a single place.

**Comment by aaronrobertshaw** [comment]
Appreciate the work on unifying these actions @jorgefilipecosta 👍 

I'll give https://github.com/WordPress/gutenberg/pull/60486 a review instead of the [previously proposed](https://github.com/WordPress/gutenberg/pull/60232#issuecomment-2039136757) follow-up.

**Pull request merged**
